src/core_tests/workflow_tasks.rs

```diff
@@ -1256,3 +1259,80 @@ async fn new_server_work_while_eviction_outstanding_doesnt_overwrite_activation(
+    let mut t = TestHistoryBuilder::default();
+    t.add_by_type(EventType::WorkflowExecutionStarted);
+    t.add_workflow_task_scheduled_and_started();
+    // Need to build the first response before adding the timeout events b/c otherwise the history
+    // builder will include them in the first task
+    let resp_1 = hist_to_poll_resp(&t, wfid.to_owned(), 1.into(), TEST_Q.to_string());
+    t.add_workflow_task_timed_out();
+    t.add_full_wf_task();
+    let timer_started_event_id = t.add_get_event_id(EventType::TimerStarted, None);
+    t.add(
+        EventType::TimerFired,
+        history_event::Attributes::TimerFiredEventAttributes(TimerFiredEventAttributes {
+            started_event_id: timer_started_event_id,
+            timer_id: "timer".to_string(),
+        }),
+    );
+    t.add_full_wf_task();
+    t.add_workflow_execution_completed();
+
+    let mut tasks = VecDeque::from(vec![resp_1]);
+    // Extend the task list with the now timeout-included version of the task. We add a bunch of
+    // them because the poll loop will spin while new tasks are available and it is buffering them
+    tasks.extend(
+        std::iter::repeat(hist_to_poll_resp(
+            &t,
+            wfid.to_owned(),
+            2.into(),
+            TEST_Q.to_string(),
+        ))
+        .take(100),
+    );
+    let mut mock = MockServerGatewayApis::new();
+    mock.expect_complete_workflow_task()
+        .returning(|_| Ok(RespondWorkflowTaskCompletedResponse::default()));
+    let mock = MocksHolder::from_gateway_with_responses(mock, tasks, vec![].into());
+    // Cache on to avoid being super repetitive
+    let mut opts = CoreInitOptionsBuilder::default();
+    opts.max_cached_workflows(10_usize);
+    let core = &mock_core_with_opts(mock, opts);
+
+    // Poll for first WFT
+    let act1 = dbg!(core.poll_workflow_task(TEST_Q).await.unwrap());
+    let poll_fut = async move {
+        // Now poll again, which will start spinning, and buffer the next WFT with timer fired in it
+        // - it won't stop spinning until the first task is complete
+        let t = dbg!(core.poll_workflow_task(TEST_Q).await.unwrap());
+        core.complete_workflow_task(WfActivationCompletion::from_cmds(
+            vec![CompleteWorkflowExecution { result: None }.into()],
+            t.run_id,
+        ))
+        .await
+        .unwrap();
+    };
+    let complete_first = async move {
+        core.complete_workflow_task(WfActivationCompletion::from_cmd(
+            StartTimer {
+                timer_id: "timer".to_string(),
+                ..Default::default()
+            }
+            .into(),
+            act1.run_id,
+        ))
+        .await
+        .unwrap();
+        dbg!("Did first complete");
+    };
+    tokio::join!(poll_fut, complete_first, async {
+        // If the shutdown is sent too too fast, we might not have got a chance to even buffer work
+        tokio::time::sleep(Duration::from_millis(5)).await;
```

Comment:
You know my opinion on time sensitive tests like this.
We should have a better way of timing the shutdown.

Reply:
Opened https://github.com/temporalio/sdk-core/issues/162